### DISCUSSION TITLE
Welcome to MindustryLogicBangLang Discussions!

### DISCUSSION THREAD

**Discussion by A4-Tacks** [A4-Tacks/mindustry_logic_bang_lang#9] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

可以在这里交流关于Bang语言的各种想法与问题，欢迎您提出宝贵的改进意见

**Comment by A4-Tacks** [comment]
接下来的开发进度可能为

1. 增加`inline-block`, 相比`expand`, 它在编译时不会增加子块域, 这可以突破一些限制 (已完成 v0.11.1)
2. 增加文件跟踪, 对构建元数据中添加文件栈, 初始化为`<stdin>`
3. 增加`includes`, 在基于`inline-block`的基础上，使用它的`Parser`对被引入文件启动一个子分析器,
   子分析器会使用母分析器的构建元数据, 并且对目标文件使用`inline-block`的内部`Parser`对目标文件进行分析.
   然后将目标文件生成的`LogicLine::Inline`返回.

   这里使用`includes`时，路径是使用的当前文件目录的相对路径, 如果是`<stdin>`, 那么将会使用`.`

**Reply by westernat** [reply]
这个子分析器使用的构建元数据是指?

**Reply by A4-Tacks** [reply]
我这进行`parse`时，会传入元数据，它记录了一些状态，如临时变量id的推进，临时标志计数的推进等。
启用另一个`parse`但是将结果产生到同一个目标时，我们不能使其冲突，所以需要使用母分析器的元数据进行分析。

这里指的是`src/syntax.rs`中的`struct Meta`

**Comment by westernat** [comment]
我注意到你的`examples/caller.mdtlbl`中
```
const Fun_Foo = fun_Foo;
const Foo = (
    take Msg = _0; # 求值出需要传递的
    take A = _1;
    take B = _2;

    const Self = Fun_Foo;

    Self.Msg Self.A Self.B = Msg A B; # 传递值
    op Self.back @counter + 1; # 设置返回地址
    goto :fun_foo_head _; # call
);
skip _ { # 跳过这段函数
    :fun_foo_head # 用于被调用的函数头部标记
    const Self = Fun_Foo;
    take Msg = Self.Msg;
    take A = Self.A;
    take B = Self.B;

    print Msg (op $ A + B;);

    @counter = Self.back; # 返回到地址
}
take["num: " 1 2] Foo;
printflush message1;
```
这一段内容.
我在想,是否可以加一组语法糖`fun`, `call`, 使其写起来类似:
```
const Foo = (
    take Msg = _0; # 求值出需要传递的
    take A = _1;
    take B = _2;

    call[Msg A B] foo_fun # 或者foo_fun(Msg, A, B)
);
fun foo_fun_head(msg, a, b) {
    print msg (op $ a + b;);
}
take["num: " 1 2] Foo;
printflush message1;
```
语法糖有时候写起来让人很舒服不是吗?

**Reply by westernat** [reply]
啊,foo_fun_head应该写成foo_fun

**Reply by A4-Tacks** [reply]
语法糖的确是很令人舒服，但是现在还没有这个计划。

**Reply by A4-Tacks** [reply]
如果我们创建该语法糖，那么将失去灵活性。
因为call的方式很多，我更希望通过繁琐一些的精细控制来自己编写call。

**Comment by A4-Tacks** [comment]
文件跟踪和includes似乎有些遥遥无期，往近看有些个可能在接下来几个版本实现的:

1. 添加`consted-DExp`，在`DExp`前面加上`const`前缀，`const(...)`展开为大致如`(__: const __Tmp = (...); setres __Tmp;)`
   并且它会进行被跳转标签收集，就像我们在正常`const`中定义的一样。

   所以我们不用担心`consted-DExp`中编写的被跳转语句被重复展开引起标签冲突的情况了。
   并且使用`consted-DExp`时不用担心使用`take`时将`DExp`直接传入时，里面定义的标记变成传入的`const`内部标记了。
   也就是说，不用担心直接传入`consted-DExp`时，定义标记与某个值的内部标记同名了。

2. `op-expr`中添加三元表达式，虽然可以预见的是连续使用可能会有多余的无条件跳转问题
3. Tag码添加无条件跳转优化，在Tag码编译时，对于无条件跳转链进行重定向优化，这可以在一些常见情况获得提升。
   当然，这会使从逻辑编译到Tag码后编译回去得到一份略优于原逻辑的逻辑码，但是这可能造成一些涉及逻辑同步性的问题。

关于第三点，似乎需要进一步讨论。并且可能的一个结果是增加编译选项来决定此行为。
但是这会比较麻烦，同时依赖于无条件跳转链的同步问题比较离谱，感觉这个优化是基本没啥问题的。

**Reply by A4-Tacks** [reply]
接下来的进度可能比较慢，还有生活也忙起来了。

三元表达式需要对`OpExpr`进行较大的改动，所以进度会比较缓慢
Tag码的优化也在等待中
现在可能优先进行更简单的`consted-DExp`

**Comment by A4-Tacks** [comment]
现在已经有更清晰的参考手册了, 可以更加清晰的了解这门语言了.
[参考手册](https://githubfast.com/A4-Tacks/mindustry_logic_bang_lang/blob/39309e859f3187e38bbcb91da2e8fb4edf629cac/examples/reference.md)

**Comment by A4-Tacks** [comment]
match已经完成，接下来的计划可能是将常量和句柄常量的目标扩展至ValueBind，并且将其放置至全局空间，用于内向外的常量传递
在之前这个操作并没有什么意义，但是现在已经拥有条件编译，它是时候上线了
理想情况下它只应被用于全局常量、匿名句柄绑定量，否则可能会增加整个系统的复杂度，需要考虑其最终目标值是否为同一个量

**Reply by A4-Tacks** [reply]
已经完成

**Comment by A4-Tacks** [comment]
接下来可能会增加需手动声明的闭包捕获量，以减少可能的冲突。
被声明为捕获的量会在声明处进行追溯捕获，这可以改善需要在展开处注意名称查找遮蔽等窘境。

还有一个较长计划是将switch的展开改为编译时而不是构建时，使其能和常量表结合，并且可以设计参数重复等

**Reply by A4-Tacks** [reply]
闭包计划完成，接下来将扩展设计将值封装为闭包和向闭包添加捕获量的内建函数

**Reply by A4-Tacks** [reply]
> 闭包计划完成，接下来将扩展设计将值封装为闭包和向闭包添加捕获量的内建函数

添加捕获量还是算了吧，想了想不是很好

**Reply by A4-Tacks** [reply]
算了想了想这俩似乎都没必要
